Index: nginx-catap/nginx/src/http/modules/perl/nginx.xs
===================================================================
--- nginx-catap.orig/nginx/src/http/modules/perl/nginx.xs
+++ nginx-catap/nginx/src/http/modules/perl/nginx.xs
@@ -474,6 +474,13 @@ header_out(r, key, value)
         r->headers_out.content_length = header;
     }
 
+    if (header->key.len == sizeof("Content-Encoding") - 1
+        && ngx_strncasecmp(header->key.data, "Content-Encoding",
+                           sizeof("Content-Encoding") - 1) == 0)
+    {
+        r->headers_out.content_encoding = header;
+    }
+
 
 void
 filename(r)
