Index: nginx-catap/nginx/auto/lib/perl/make
===================================================================
--- nginx-catap.orig/nginx/auto/lib/perl/make
+++ nginx-catap/nginx/auto/lib/perl/make
@@ -28,6 +28,7 @@ $NGX_OBJS/src/http/modules/perl/Makefile
 			NGX_PCRE=$PCRE					\
 			NGX_OBJS=$NGX_OBJS				\
 			NGX_SRC_DIR=$NGX_SRC_DIR			\
+			NGX_OPENSSL=$OPENSSL                            \
 		$NGX_PERL Makefile.PL					\
 			LIB=$NGX_PERL_MODULES				\
 			INSTALLSITEMAN3DIR=$NGX_PERL_MODULES_MAN
Index: nginx-catap/nginx/src/http/modules/perl/Makefile.PL
===================================================================
--- nginx-catap.orig/nginx/src/http/modules/perl/Makefile.PL
+++ nginx-catap/nginx/src/http/modules/perl/Makefile.PL
@@ -24,7 +24,12 @@ WriteMakefile(
                          "-I $ENV{NGX_OBJS} " .
                          ($ENV{NGX_PCRE} =~ /^(YES|NO)/ ? "" :
                              ($ENV{NGX_PCRE} =~ m#^/# ? "-I $ENV{NGX_PCRE} " :
-                                  "-I $ENV{NGX_SRC_DIR}/$ENV{NGX_PCRE} ")),
+                                  "-I $ENV{NGX_SRC_DIR}/$ENV{NGX_PCRE} ")) .
+                         ($ENV{NGX_OPENSSL} =~ /^(YES|NO)/ ? "" :
+                             ($ENV{NGX_OPENSSL} =~ m#^/# ?
+                                  "-I $ENV{NGX_OPENSSL}/.openssl/include " :
+                      "-I $ENV{NGX_SRC_DIR}/$ENV{NGX_OPENSSL}/.openssl/include ")),
+
 
     depend => {
         'nginx.c'     =>
