Index: nginx-catap/nginx/src/http/ngx_http_upstream_round_robin.c
===================================================================
--- nginx-catap.orig/nginx/src/http/ngx_http_upstream_round_robin.c
+++ nginx-catap/nginx/src/http/ngx_http_upstream_round_robin.c
@@ -667,7 +667,7 @@ failed:
 static ngx_uint_t
 ngx_http_upstream_get_peer(ngx_http_upstream_rr_peers_t *peers)
 {
-    ngx_uint_t                    i, n;
+    ngx_uint_t                    i, n, reset = 0;
     ngx_http_upstream_rr_peer_t  *peer;
 
     peer = &peers->peer[0];
@@ -706,6 +706,10 @@ ngx_http_upstream_get_peer(ngx_http_upst
             return n;
         }
 
+        if (reset++) {
+            return 0;
+        }
+
         for (i = 0; i < peers->number; i++) {
             peer[i].current_weight = peer[i].weight;
         }
